integrations/wiz/client.py

```diff
@@ -0,0 +1,246 @@
+        externalId
+        name
+        url
+      }
+      notes {
+        createdAt
+        updatedAt
+        text
+        user {
+          name
+          email
+        }
+        serviceAccount {
+          name
+        }
+      }
+    }
+    pageInfo {
+      hasNextPage
+      endCursor
+    }
+  }
+}
+"""
+
+
+class TokenResponse(BaseModel):
+    access_token: str = Field(alias="access_token")
+    expires_in: int = Field(alias="expires_in")
+    token_type: str = Field(alias="token_type")
+    _retrieved_time: int = PrivateAttr(get_time())
+
+    @property
+    def expired(self) -> bool:
+        return self._retrieved_time + self.expires_in < get_time()
+
+    @property
+    def full_token(self) -> str:
+        return f"{self.token_type} {self.access_token}"
+
+
+class CacheKeys(StrEnum):
+    ISSUES = "wiz_issues"
+
+
+class WizClient:
+    def __init__(
+        self, api_url: str, client_id: str, client_secret: str, token_url: str
+    ):
+        self.api_url = api_url
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.token_url = token_url
+        self.last_token_object: TokenResponse | None = None
+
+        self.http_client = http_async_client
+
+    @property
+    async def api_auth_headers(self) -> dict[str, Any]:
+        token = self._get_token()
+
+        return {
+            "Authorization": token.full_token,
+            "Content-Type": "application/json",
+        }
+
+    @property
+    def auth_params(self) -> dict[str, Any]:
+        audience_mapping = {
+            **{url: "beyond-api" for url in AUTH0_URLS},
+            **{url: "wiz-api" for url in COGNITO_URLS},
+        }
+
+        if self.token_url not in audience_mapping:
+            raise Exception("Invalid Token URL")
+
+        audience = audience_mapping[self.token_url]
+
+        return {
+            "grant_type": "client_credentials",
+            "audience": audience,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+
+    def _get_token(self) -> TokenResponse:
+        try:
+            response = httpx.post(
+                self.token_url,
+                data=self.auth_params,
+                headers={"Content-Type": "application/x-www-form-urlencoded"},
+            )
+            response.raise_for_status()
+            return TokenResponse(**response.json())
+        except Exception as e:
+            logger.exception(e)
+            raise
+
+    @property
+    async def token(self) -> str:
+        if not self.last_token_object or not self.last_token_object.is_valid:
+            msg = "Wiz Token expired or invalid, fetching new token"
+            logger.info(msg)
+            self.last_token_object = await self._get_token(
+                self.client_id, self.client_secret
+            )
+
+        return self.last_token_object.token
+
+    async def get_issues(self) -> AsyncGenerator[list[dict[str, Any]], None]:
```

Comment:
let me know what u think

We should either seperate the issues query from the projects or yield just unique projects because one project may have multiple issues which means that if i have 2000 issues then i will also try to resync the same project 2000 times

Reply:
Oh! Yes, on a large scale, it becomes undesirable. I have asked @MPTG94 to help me get the `read project` scope added to the credentials I am using. We'll use the projects endpoint and separate them out.